sc-memory/sc-memory/sc_agent_manager.tpp

```diff
@@ -135,57 +154,153 @@ void ScAgentManager<TScAgent>::Unsubscribe(
+    if (!subscription)
       SC_THROW_EXCEPTION(
           utils::ExceptionInvalidState,
-          "Agent `" << agentName << "` does not have been subscribed to event `" << eventClassName << "("
+          "Agent `" << agentClassName << "` does not have been subscribed to event `" << eventClassName << "("
                     << subscriptionElementAddr.Hash() << ")` yet.");
 
     SC_LOG_INFO(
-        "Unsubscribe agent `" << agentName << "` from event `" << eventClassName << "("
+        "Unsubscribe agent `" << agentClassName << "` from event `" << eventClassName << "("
                               << subscriptionElementAddr.Hash() << ")`.");
 
-    delete it->second;
-    subscriptionsMap.erase(subscriptionElementAddr);
+    delete subscription;
+    EraseSubscription(subscriptionElementAddr, subscriptions);
   }
 
-  if (subscriptionsMap.empty())
-    ScAgentManager<TScAgent>::m_agentSubscriptions.erase(agentName);
+  ClearEmptyAgentImplementationSubscriptions(
+      agentClassName, agentImplementationAddr, agentImplementationsToSubscriptions, subscriptions);
+}
+
+template <class TScAgent>
+bool ScAgentManager<TScAgent>::IsAgentSubscribedToEventOfErasingElement(std::string const & agentClassName)
+{
+  auto const & eventClassIt = ScAgentManager<TScAgent>::m_agentEventClasses.find(agentClassName);
+  return eventClassIt != ScAgentManager<TScAgent>::m_agentEventClasses.cend()
+         && eventClassIt->second == ScKeynodes::sc_event_before_erase_element;
+}
+
+template <class TScAgent>
+typename ScAgentManager<TScAgent>::ScAgentImplementationsToSubscriptionsRef ScAgentManager<
+    TScAgent>::ResolveAgentClassAgentImplementationSubscriptions(std::string const & agentClassName)
+{
+  if (!ScAgentManager<TScAgent>::m_agentClassesToAgentImplementationSubscriptions.count(agentClassName))
+    ScAgentManager<TScAgent>::m_agentClassesToAgentImplementationSubscriptions.insert({agentClassName, {}});
+
+  return GetAgentClassAgentImplementationSubscriptions(agentClassName);
+}
+
+template <class TScAgent>
+typename ScAgentManager<TScAgent>::ScAgentImplementationsToSubscriptionsRef ScAgentManager<
+    TScAgent>::GetAgentClassAgentImplementationSubscriptions(std::string const & agentClassName)
+{
+  auto const & agentImplementationsToSubscriptionsIt =
+      ScAgentManager<TScAgent>::m_agentClassesToAgentImplementationSubscriptions.find(agentClassName);
+  if (agentImplementationsToSubscriptionsIt
+      == ScAgentManager<TScAgent>::m_agentClassesToAgentImplementationSubscriptions.cend())
+    return std::nullopt;
+
+  return std::ref(agentImplementationsToSubscriptionsIt->second);
+}
+
+template <class TScAgent>
+typename ScAgentManager<TScAgent>::ScSubscriptionsRef ScAgentManager<TScAgent>::
+    GenerateAgentImplementationSubscriptions(
+        ScAddr const & agentImplementationAddr,
+        ScAgentManager<TScAgent>::ScAgentImplementationsToSubscriptionsRef & agentImplementationsToSubscriptions)
```

Comment:
these methods have std::optional as their arguments but they don't check its validity

Reply:
It will be redundant. They are protected. SubscribeAgent method creates them, works with them and destroys them. Users don't know about them. Nobody don't know about them. I can make them private, if you wish

Reply:
why is it saying that optional is passed if optional validity is checked before calls and they can receive just what's inside of the optional?

Reply:
Fixed